solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/outbound/JCSMPOutboundMessageHandler.java

```diff
@@ -96,28 +102,64 @@ public void handleMessage(Message<?> message) throws MessagingException {
 					String.format("Unable to parse header %s", SolaceBinderHeaders.CONFIRM_CORRELATION), e);
 		}
 
-		XMLMessage xmlMessage = xmlMessageMapper.map(message, properties.getExtension().getHeaderExclusions(),
-				properties.getExtension().isNonserializableHeaderConvertToString());
-		correlationKey.setRawMessage(xmlMessage);
-		xmlMessage.setCorrelationKey(correlationKey);
+		List<XMLMessage> smfMessages;
+		Object proxyCorrelationKey;
+		if (message.getHeaders().containsKey(SolaceBinderHeaders.BATCHED_HEADERS)) {
+			LOGGER.debug("Detected header {}, handling as batched message (Message<List<?>>) <message handler ID: {}>",
+					SolaceBinderHeaders.BATCHED_HEADERS, id);
+			smfMessages = xmlMessageMapper.mapBatchMessage(
+					message,
+					properties.getExtension().getHeaderExclusions(),
+					properties.getExtension().isNonserializableHeaderConvertToString());
 
-		LOGGER.debug("Publishing message to destination [ {}:{} ]",
-				targetDestination instanceof Topic ? "TOPIC" : "QUEUE", targetDestination);
+			proxyCorrelationKey = transactedSession != null ?
+					correlationKey : new BatchProxyCorrelationKey(correlationKey, smfMessages.size());
+		} else {
+			smfMessages = List.of(xmlMessageMapper.map(
+					message,
+					properties.getExtension().getHeaderExclusions(),
+					properties.getExtension().isNonserializableHeaderConvertToString()));
+			proxyCorrelationKey = correlationKey;
+		}
+
+		correlationKey.setRawMessages(smfMessages);
 
 		try {
-			producer.send(xmlMessage, targetDestination);
+			for (int i = 0; i < smfMessages.size(); i++) {
+				XMLMessage smfMessage = smfMessages.get(i);
+				smfMessage.setCorrelationKey(proxyCorrelationKey);
+
+				LOGGER.debug("Publishing message {} of {} to destination [ {}:{} ] <message handler ID: {}>",
+						i + 1, smfMessages.size(), targetDestination instanceof Topic ? "TOPIC" : "QUEUE",
+						targetDestination, id);
+
+				producer.send(smfMessage, targetDestination);
```

Comment:
Can each message have its own dynamic destination? It looks like `targetDestination` applies for the whole batch? 
Typically, I believe a dynamic destination is likely to be derived from the internals of a message.

Reply:
Yeah, `targetDestination` and `solace_binder_targetDestinationType` should be set per-message. Not at the batched message level. I'll fix that.